### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by LeoTindall**
This converts the whole project to Typescript and fixes a few minor errors discovered by the typechecker.

### CODE REVIEW

File: src/client/app/actions/graph.ts [hunk 1]
```diff
@@ -71,7 +71,7 @@ function shouldChangeGraphZoom(state, timeInterval) {
 
 export function changeGraphZoomIfNeeded(timeInterval) {
 	return (dispatch, getState) => {
-		if (shouldChangeGraphZoom(getState())) {
+		if (shouldChangeGraphZoom(getState(), "all")) {
```
Comment (LeoTindall) [inline src/client/app/actions/graph.ts]: This is likely not correct, as it seems to cause the following error in the server:

```
GET /api/meters 200 28.757 ms - 396
Deprecation warning: value provided is not in a recognized RFC2822 or ISO format. moment construction falls back to js Date(), which is not reliable across all browsers and versions. Non RFC2822/ISO date formats are discouraged and will be removed in an upcoming major release. Please refer to http://momentjs.com/guides/#/warnings/js-date/ for more info.
Arguments:
[0] _isAMomentObject: true, _isUTC: false, _useUTC: false, _l: undefined, _i: 4, _f: undefined, _strict: undefined, _locale: [object Object]
Error
    at Function.createFromInputFallback (/usr/src/app/node_modules/moment/moment.js:324:94)
    at configFromString (/usr/src/app/node_modules/moment/moment.js:2363:11)
    at configFromInput (/usr/src/app/node_modules/moment/moment.js:2589:9)
    at prepareConfig (/usr/src/app/node_modules/moment/moment.js:2572:9)
    at createFromConfig (/usr/src/app/node_modules/moment/moment.js:2539:40)
    at createLocalOrUTC (/usr/src/app/node_modules/moment/moment.js:2626:12)
    at createLocal (/usr/src/app/node_modules/moment/moment.js:2630:12)
    at hooks (/usr/src/app/node_modules/moment/moment.js:16:25)
    at stringified.split.map.timestamp (/usr/src/app/src/common/TimeInterval.js:55:82)
    at Array.map (<anonymous>)
```

File: package.json [hunk 2]
```diff
@@ -58,6 +56,10 @@
     "xml2js": "~0.4.19"
   },
   "devDependencies": {
+    "@types/core-js": "~0.9.43",
```
Comment (johndiiorio) [inline package.json]: Why are some types in dev, some not in dev? Shouldn't they all be in dev?

File: package.json [hunk 3]
```diff
@@ -70,7 +72,9 @@
     "mocha": "~3.5.3",
     "mock-require": "~2.0.2",
     "redux-devtools": "~3.4.0",
-    "sinon": "~4.0.2",
-    "style-loader": "~0.19.0"
+    "source-map-loader": "~0.2.3",
+    "style-loader": "~0.19.0",
+    "typescript": "~2.6.1",
```
Comment (johndiiorio) [inline package.json]: Not in alphabetical order - don't manually edit package.json by hand, use npm install to do so

File: src/client/app/components/LoginComponent.tsx [hunk 4]
```diff
@@ -2,12 +2,26 @@
...
 
-import React from 'react';
+import * as React from 'react';
 import axios from 'axios';
 import { browserHistory } from 'react-router';
 import HeaderComponent from '../components/HeaderComponent';
 
-export default class LoginComponent extends React.Component {
+interface NotificationSpec {
+	message: string, level: string, position: string, autoDismiss: number
```
Comment (johndiiorio) [inline src/client/app/components/LoginComponent.tsx]: Better to have these on multiple lines.
There also might be additional parameters that can be passed to ReactNotificationSystem.

File: src/client/app/containers/BarChartContainer.ts [hunk 5]
```diff
@@ -36,8 +36,11 @@ function mapStateToProps(state) {
 			}
 		}
 	}
-    // Converts the label set into an array for Chart.js and sorts the labels based on the first date of the time interval
-	data.labels = Array.from(labelsSet).sort((x, y) => moment(x.split(' - ')[0], 'MMM DD, YYYY').format('x') - moment(y.split(' - ')[0], 'MMM DD, YYYY').format('x'));
+	// Converts the label set into an array for Chart.js and sorts the labels based on the first date of the time interval
+	data.labels = Array.from(labelsSet).sort((x, y) => {
+		const t1 = moment(x.split(' - ')[0], 'MMM DD, YYYY').format('x');
+		const t2 = moment(y.split(' - ')[0], 'MMM DD, YYYY').format('x');
+		return +(t1) - +(t2);});
```
Comment (johndiiorio) [inline src/client/app/containers/BarChartContainer.ts]: You should move the }); to a new line

File: package.json (hunk 3 above)
Comment (LeoTindall) [reply package.json]: There was a merge conflict, sorry.

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+ 
+{
+    "compilerOptions": {
+        "baseUrl": "./src/client/app/",
+        "outDir": "./src/client/public/",
+        "sourceMap": true,
+        "noImplicitAny": false,
```
Comment (johndiiorio) [inline tsconfig.json]: This should be set to `true`

File: src/client/index.html [hunk 7]
```diff
@@ -11,6 +11,6 @@
 </head>
 <body>
     <div id="root"></div>
-	<script src="public/bundle.js" type="text/javascript"></script>
+	<script src="public/dist/bundle.js" type="text/javascript"></script>
```
Comment (johndiiorio) [inline src/client/index.html]: Why did you change this to /dist?

**Comment by huss** [comment]
Somehow this message did not make it into this thread via email so I am posting now. It is from yesterday morning:
While I think this is valuable work that we planned to integrate soon, I thought the resolution of the discussion yesterday was this would not be done until after groups and any other near done work is in the main branch and labeled. That will be an official release/label. Then, the new work such as this will be a branch/commit from that starting point and become part of the next release. That may be the plan or maybe I have a wrong recollection of the meetings but I wanted to put this out there.
